src/type2.c

```diff
@@ -37,11 +38,15 @@ SEXP vctrs_type2(SEXP x, SEXP y) {
 
   default:
   dispatch:
-    return vctrs_dispatch3(vec_type2_dispatch_fn, x, y);
+    return vctrs_dispatch2(syms_vec_type2_dispatch, fns_vec_type2_dispatch,
+                           syms_x, x,
+                           syms_y, y,
+                           R_GlobalEnv);
```

Comment:
Why is this environment different to `vec_slice()` dispatch?

Reply:
Because we don't forward dots so we don't need to dispatch in the exec env of `vec_type2()`. This dispatches in a child of the global env without needing to add parameters and pass an environment across the C functions.

Maybe this should be `R_NilValue` so it doesn't suggest dispatch occurs exactly in the global env rather than a mask inheriting from it.

Reply:
Dispatch in the global environment is now signalled by `R_NilValue`, otherwise dispatch occurs in the environment supplied.

Reply:
Isn't it better to be consistent across all our internal generics?

Reply:
I don't think it really matters in this case. Ideally we wouldn't dispatch in the vctrs namespace at all, we'd create a child of the global env and forward dots there. The PR implementation makes things a bit simpler. Would you like to forward dots in a child of global instead?

Reply:
It just worries me that we have 3 (?) internal generics and two dispatch methods.

Reply:
I will change it to always dispatch in a child of the global environment.